src/libraries/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -1645,26 +1645,46 @@ private static XmlSerializationCollectionFixupCallback GetCreateCollectionOfObje
 
                     if (member.Source == null)
                     {
+                        var isList = mapping.TypeDesc.IsArrayLike && !mapping.TypeDesc.IsArray;
                         var pi = member.Mapping.MemberInfo as PropertyInfo;
-                        if (pi != null && typeof(IList).IsAssignableFrom(pi.PropertyType)
-                            && (pi.SetMethod == null || !pi.SetMethod.IsPublic))
+
+                        // Old serializers would trip over private property setters. Except for lists, which get special treatment.
+                        // So do the same here, even though this serializer is more capable of handling private setters. *sigh*
+                        // Do note that this still allows Lists with private setters to utilize that capability of a reflection serializer.
+                        if (!isList && pi != null && pi.SetMethod != null && !pi.SetMethod.IsPublic)
                         {
-                            member.Source = (value) =>
+                            member.Source = (value) => throw new InvalidOperationException(SR.Format(SR.XmlReadOnlyPropertyError, pi.Name, pi.DeclaringType!.FullName));
+                        }
+
+                        // If a list doesn't have a setter, or only has a private setter, we shouldn't just ignore it, because the default
+                        // constructor for the type might have created a list we can use.
```

Comment:
Is this comment correct? If a list property has a private setter, it will execute the code in the first if block and won't hit this else if expression.

Reply:
The `if` condition applies to non-list properties. This `else if` applies to list properties. So a list-like property with a private setter will still end up here. This whole section is ugly and dumb, and if it was a greenfield serializer I'd probably consider the behaviors here wrong. But the ILGen serializer does these things. Thus all the new tests to tease out some of the odd corner-case behaviors that we weren't necessarily testing before, so we can have more confidence that the reflection serializer matches behaviors.

Reply:
Comment structure updated.